internal/planner/operation.go

```diff
@@ -0,0 +1,174 @@
+	isDone       bool
+}
+
+func (n *operationNode) Spans(spans core.Spans) {
+	for _, child := range n.children {
+		child.Spans(spans)
+	}
+}
+
+func (n *operationNode) Kind() string {
+	return operationNodeKind
+}
+
+func (n *operationNode) Init() error {
+	n.isDone = false
+	n.currentValue = core.Doc{}
+
+	for _, child := range n.children {
+		err := child.Init()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (n *operationNode) Start() error {
+	for _, child := range n.children {
+		err := child.Start()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (n *operationNode) Close() error {
+	for _, child := range n.children {
+		err := child.Close()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (n *operationNode) Source() planNode {
+	return nil
+}
+
+func (p *operationNode) Children() []planNode {
+	return p.children
+}
+
+func (n *operationNode) Next() (bool, error) {
+	if n.isDone {
+		return false, nil
+	}
+
+	n.currentValue = n.documentMapping.NewDoc()
+	for i, child := range n.children {
+		switch child.(type) {
+		case *topLevelNode:
+			hasChild, err := child.Next()
+			if err != nil {
+				return false, err
+			}
+			if !hasChild {
+				return false, ErrMissingChildValue
+			}
+			n.currentValue.Fields[n.childIndexes[i]] = child.Value().Fields[0]
+
+		default:
+			var docs []core.Doc
+			for {
+				hasChild, err := child.Next()
+				if err != nil {
+					return false, err
+				}
+				if !hasChild {
+					break
+				}
+				docs = append(docs, child.Value())
+			}
+			n.currentValue.Fields[n.childIndexes[i]] = docs
+		}
+	}
+
+	n.isDone = true
+	return true, nil
+}
+
+// Operation creates a new operationNode using the given Selects.
+func (p *Planner) Operation(operation *mapper.Operation) (*operationNode, error) {
+	var children []planNode
+	var childIndexes []int
+
+	for _, s := range operation.Selects {
+		if _, isAgg := request.Aggregates[s.Name]; isAgg {
+			// If this Select is an aggregate, then it must be a top-level
+			// aggregate and we need to resolve it within the context of a
+			// top-level node.
+			child, err := p.Top(s)
+			if err != nil {
+				return nil, err
+			}
+			children = append(children, child)
+			childIndexes = append(childIndexes, s.Index)
+		} else {
+			child, err := p.Select(s)
+			if err != nil {
+				return nil, err
+			}
+			children = append(children, child)
+			childIndexes = append(childIndexes, s.Index)
+		}
+	}
+
+	for _, m := range operation.Mutations {
+		child, err := p.newObjectMutationPlan(m)
+		if err != nil {
+			return nil, err
+		}
+		children = append(children, child)
+		childIndexes = append(childIndexes, m.Index)
+	}
+
+	for _, s := range operation.CommitSelects {
+		child, err := p.CommitSelect(s)
+		if err != nil {
+			return nil, err
+		}
+		children = append(children, child)
+		childIndexes = append(childIndexes, s.Index)
+	}
+
+	if len(children) == 0 {
+		return nil, ErrOperationDefinitionMissingSelection
```

Comment:
question: Is this spec-compliant? A selection set must always be provided?

Reply:
I didn't see anything in the spec about empty selections. I've removed this error since it should be fine to return empty results.